### PR TITLE
[AutoScheduler] Improve the rule of mutating parallel granularity

### PR DESCRIPTION
**Pull request by merrymercy**
- Improve the`MutateParallel` in the evolutionary search
  - The old rule will only try to fuse and parallel more for loops. This is because the condition in this line (https://github.com/apache/incubator-tvm/blob/1d5504bfeabaa8f4c64115c1e0624991cfdbca98/src/auto_scheduler/search_policy/sketch_policy_rules.cc#L1112) is wrong. `it->annotation != IteratorAnnotation::kNone` is always true. (`it->annotation` is always `IteratorAnnotation::kParallel`)
  - The old rule does not handle `PragmaStep`
  - The new rule fixed the above limitations of old rules.
- Add a test case for `MutateParallel`

### CODE REVIEW

File: src/auto_scheduler/search_policy/sketch_policy_rules.cc [hunk 1]
```diff
@@ -1081,52 +1085,43 @@ PopulationGenerationRule::ResultKind MutateParallel::Apply(SketchPolicyNode* pol
...
-  auto fused_ids = fuse_step->fused_ids;
-  std::vector<double> fuse_dir = {0.5, 1.0};
-
-  // The case that we can only fuse more. This may happen after multiple mutations.
-  if (fused_ids.size() == 1) {
-    fuse_dir[0] = 0.0;
-  }
+  int stage_id = fuse_step->stage_id;
+  const Stage& stage = tmp_s->stages[stage_id];
+  size_t iter_id;
```
Comment (comaniac) [inline src/auto_scheduler/search_policy/sketch_policy_rules.cc:1101]: Maybe call it `max_fusable_iter_id` or something similar to make it clearer?

File: tests/python/unittest/test_auto_scheduler_evolutionary_search.py [hunk 2]
```diff
@@ -22,56 +22,102 @@
...
+    policy = auto_scheduler.SketchPolicy(task, program_cost_model=MockCostModel(), verbose=0)
+    states = policy.sample_initial_population(100)
+
+    bad_states = []
+    for state in states:
+        if not MockCostModel.is_good_state(state):
+            bad_states.append(state)
+
+    found = False
+    retry_ct = 0
```
Comment (comaniac) [inline tests/python/unittest/test_auto_scheduler_evolutionary_search.py:111]: Where does this get updated?

**Review by comaniac — COMMENTED** [review]
LGTM. Thanks! The refined approach looks more concise and effective. Just some nits.

File: tests/python/unittest/test_auto_scheduler_evolutionary_search.py (hunk 2 above)
Comment (merrymercy) [reply tests/python/unittest/test_auto_scheduler_evolutionary_search.py:108]: Good catch. Fixed.

**Review by comaniac — APPROVED** [review]

**Comment by zhiics** [comment]
@merrymercy @comaniac @FrozenGene it looks test_mutate_parallel fails in the master CI. Could you guys take a look?

**Comment by comaniac** [comment]
#6572

**Comment by merrymercy** [comment]
Besides the flaky test, I also found this mutation rule leads to occasional segment fault/ floating-point exception during long auto-tuning.
I will investigate and send the fix soon.

**Comment by merrymercy** [comment]
Fixed by #6657

**Pull request merged**